opensearch/src/main/java/org/opensearch/sql/opensearch/storage/scan/CalciteLogicalIndexScan.java

```diff
@@ -189,4 +246,155 @@ public CalciteLogicalIndexScan pushDownLimit(Integer limit, Integer offset) {
+      List<RelFieldCollation> mergedCollations =
+          mergeCollations(existingFieldCollations, collations);
+
+      // Propagate the sort to the new scan
+      RelTraitSet traitsWithCollations = getTraitSet().plus(RelCollations.of(mergedCollations));
+      CalciteLogicalIndexScan newScan =
+          new CalciteLogicalIndexScan(
+              getCluster(),
+              traitsWithCollations,
+              hints,
+              table,
+              osIndex,
+              getRowType(),
+              cloneWithoutSort(pushDownContext));
+
+      List<SortBuilder<?>> builders = new ArrayList<>();
+      for (RelFieldCollation collation : mergedCollations) {
+        int index = collation.getFieldIndex();
+        String fieldName = this.getRowType().getFieldNames().get(index);
+        RelFieldCollation.Direction direction = collation.getDirection();
+        RelFieldCollation.NullDirection nullDirection = collation.nullDirection;
+        // Default sort order is ASCENDING
+        SortOrder order =
+            RelFieldCollation.Direction.DESCENDING.equals(direction)
+                ? SortOrder.DESC
+                : SortOrder.ASC;
+        // TODO: support script sort and distance sort
+        SortBuilder<?> sortBuilder;
+        if (ScoreSortBuilder.NAME.equals(fieldName)) {
+          sortBuilder = SortBuilders.scoreSort();
+        } else {
+          String missing =
+              switch (nullDirection) {
+                case FIRST -> "_first";
+                case LAST -> "_last";
+                default -> null;
+              };
+          // Keyword field is optimized for sorting in OpenSearch
+          String fieldNameKeyword =
+              OpenSearchTextType.convertTextToKeyword(
+                  fieldName, osIndex.getFieldTypes().get(fieldName));
+          sortBuilder = SortBuilders.fieldSort(fieldNameKeyword).missing(missing);
+        }
+        builders.add(sortBuilder.order(order));
+      }
+      newScan.pushDownContext.add(
+          PushDownAction.of(
+              PushDownType.SORT,
+              builders.toString(),
+              requestBuilder -> requestBuilder.pushDownSort(builders)));
+      return newScan;
+    } catch (Exception e) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Cannot pushdown the sort {}", getCollationNames(collations), e);
+      } else {
+        LOG.warn("Cannot pushdown the sort {}, ", getCollationNames(collations));
+      }
+    }
+    return null;
+  }
+
+  private List<String> getCollationNames(List<RelFieldCollation> collations) {
+    return collations.stream()
+        .map(collation -> getRowType().getFieldNames().get(collation.getFieldIndex()))
+        .toList();
+  }
+
+  /**
+   * Check if the sort by collations contains any aggregators that are pushed down. E.g. In `stats
+   * avg(age) as avg_age by state | sort avg_age`, the sort clause has `avg_age` which is an
```

Comment:
it may have a more complex case:
```
avg(age) as avg_age by state | rename avg_age as new_avg_age | sort new_avg_age
```
I don't think the current `hasAggregatorInSortBy` works for that.